### PR TITLE
[PYG-181] 💥Support recursive delete

### PR DESCRIPTION
**Pull request by doctrino**
## Description
Please describe the change you have made.

## Checklist:
- [ ] Tests added/updated.
- [ ] Documentation updated.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-gql-pygen/blob/main/CHANGELOG.md).
- [ ] Version bumped. If triggering a new release is desired, bump the version number in
  [version.py](https://github.com/cognitedata/cognite-gql-pygen/blob/main/cognite/gqlpygen/version.py) and
  [pyproject.toml](https://github.com/cognitedata/cognite-gql-pygen/blob/main/pyproject.toml) per [semantic versioning](https://semver.org/).
- [ ] Regenerate example SDK `export PYTHONPATH=. && python dev.py generate`. Need to be run both
  for `pydantic` `v1` and `v2` environments.

### CODE REVIEW

File: tests/test_integration/test_delete.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+if IS_PYDANTIC_V2:
+    from omni import OmniClient
+    from omni import data_classes as dc
+else:
+    from omni_pydantic_v1 import OmniClient
+    from omni_pydantic_v1 import data_classes as dc
+
+
+def test_delete_recursive(omni_client: OmniClient, cognite_client: CogniteClient) -> None:
```
Comment (doctrino) [inline tests/test_integration/test_delete.py:15]: This test demonstrates the new behavior @Juliamg

Comment (Juliamg) [reply tests/test_integration/test_delete.py:15]: The test looks good and demonstrates the functionality of deleting the connecting node and the edge that connects them.

What will happen in the case where Item B has an edge to another node, say Item C. Would the deletion of Item A still delete item B, or is there a check somewhere that item B has no other edges to any other nodes, which would mean that the delete operation should keep Item B, delete Item A and the edge between Item A and Item B?

File: cognite/pygen/_core/templates/_api_client_multi_model.py.jinja [hunk 2]
```diff
@@ -154,12 +163,19 @@ class {{ client_name }}:
         return self.upsert(items, replace, write_none)
 
     def delete(
-        self, external_id: str | SequenceNotStr[str], space: str = DEFAULT_INSTANCE_SPACE
+        self,
+        external_id: (
+            str | SequenceNotStr[str] | data_classes.DomainModelWrite | Sequence[data_classes.DomainModelWrite]
+        ),
+        space: str = DEFAULT_INSTANCE_SPACE
```
Comment (Juliamg) [inline cognite/pygen/_core/templates/_api_client_multi_model.py.jinja:170]: Could it be an idea to add a depth parameter to specify the depth for which the delete operation should happen? 

We might not always want to delete the connecting node as well in the case where that node has edges to other nodes again, but we would always want to delete the edges (this is the scenario I describe in the comment to the test below).

So the default depth could be 0 - that upon deletion of a node, the edges going in and out of that node are deleted, and then you could provide depth=1, if you in addition want to delete the connecting nodes one level out, and possibly even 2 or 3?

File: tests/test_integration/test_delete.py (hunk 1 above)
Comment (doctrino) [reply tests/test_integration/test_delete.py:15]: Yes, that is correct. The delete call will only delete the nodes and edges that are passed in. If there exists other edges or connected nodes, there is no-lookup to delete them.

File: cognite/pygen/_core/templates/_api_client_multi_model.py.jinja (hunk 2 above)
Comment (doctrino) [reply cognite/pygen/_core/templates/_api_client_multi_model.py.jinja:170]: Hmm, would have to be connection hops or something as depth is a property of a tree not a graph. 

But I am slightly skeptical about it as it complicates the call. Even with a good default.

**Review by Juliamg — APPROVED** [review]

**Pull request merged**